hasura/graphql-engine
# Add MonadTrace and MonadExecuteQuery abstractions

### Description

A combination of two changes:

* Adding `MonadExecuteQuery` to support switching out the code we use to actually execute compiled SQL code
* Adding `MonadTrace` and `HasReporter` to support tracing with a flexible tracing backend implementation

### Changelog


### Affected components


### Related Issues

* TODO

### Solution and Design

Tracing interfaces and implementation are provided for adding the ability to perform fine-grained tracing throughout actions as they are performed by the application. Implementation can also be deferred so that library-use of GraphQL-Engine can implement different tracing mechanisms as required.

* The addition of two new classes in `Hasura.Tracing` - `HasReporter` and `MonadTrace` provide an interface for tracing actions
* An implementation via `TraceT` to allow for tracing to be performed

Caching foundations have been implemented to allow for data cahcing of queries and actions.

### Steps to test and verify

* Existing test-suites and integration testing should be performed to ensure that this has not introduced any regressions as current behaviour of the application should remain the same
* No impact on the user-interfaces in graphql-engine - Downstream codebases should be tested to verify that custom tracing implementations are executed as expected.

#### Catalog upgrade

Does this PR change Hasura Catalog version?

#### Metadata

Does this PR add a new Metadata feature?


#### GraphQL

#### Breaking changes
